modules/assignment/src/main/java/com/intuit/wasabi/assignment/impl/AssignmentsImpl.java

```diff
@@ -963,24 +939,46 @@ private Assignment generateAssignment(Experiment experiment, User.ID userID, Con
         Assignment result = builder.build();
         return assignmentsRepository.assignUser(result, experiment, date);
     }
-
+    
     /**
      * Returns a bucketList depending on whether it is a personalization experiment or not.
-     * Retrieves buckets from DE if personalization is enabled
+     * Retrieves buckets from Assignment Decorator if personalization is enabled
      * Retrieves buckets from Repository if personalization is not enabled and if skipBucketRetrieval is false
      *
      * @param experiment          Experiment
+     * @param userID              UserId
+     * @param segmentationProfile The segmentation profile
      * @param skipBucketRetrieval If false, retrieves buckets from Repository
      * @return BucketList
      */
-    protected BucketList getBucketList(Experiment experiment, Boolean skipBucketRetrieval) {
-        BucketList buckets = null;
-        if (!skipBucketRetrieval) {
-            buckets = repository.getBuckets(experiment.getID());
+    protected BucketList getBucketList(Experiment experiment, User.ID userID, SegmentationProfile segmentationProfile, Boolean skipBucketRetrieval) {
+        
+        BucketList bucketList = null;
+        Boolean isPersonalizationEnabled = experiment.getIsPersonalizationEnabled();
+        if (isPersonalizationEnabled && !Objects.isNull(assignmentDecorator)) {
+            try {
+                bucketList = assignmentDecorator.getBucketList(experiment, userID, segmentationProfile);
+            } catch (BucketDistributionNotFetchableException | IOException e) {
+                LOGGER.error(new StringBuilder("Error obtaining the Bucket List from ").append(assignmentDecorator.getClass().getSimpleName())
```

Comment:
@bargenilesh going to eliminate IOException altogether from AssignmentDecorator.getBucketList() interface. This method should not throw IOException. It should be abstract out from the actual implementation of this method by sub class.